### PR TITLE
Ofira jwt OIDC header1

### PR DESCRIPTION
**Pull request by oburstein-hub**
### Desired Outcome

Adding capability to do OIDC authentication when the ID Token is passed in header

### Implemented Changes

The changes are done in controller and its helping classes relevant to OIDC
The changes are to handle requests that get ID Token in header instead of body of request
The request that comes without IDToken in body is trying to read Authorization token from header and 
checks if it has a "Bearer " prefix. After it the prefix is cut-out and the token after it is passed as an ID Toke
to the existing validation mechanism

### Connected Issue/Story

Resolves [ONYX-23732](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-23732)

### Definition of Done
OIDC authentication for UI is implemented and passed first testing

#### Changelog

- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [x] This PR does not require updating any documentation

#### Behavior

- [ x] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (rafis3) [inline app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:15]: Isn't the `request` and `credentials` overlapping? The `credentials` value is supposed to be taken from the body of the request, but then we send the entire request as another parameter, which contains the body as well.

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:48]: Why does the test need this new content type? Doesn't that mean the API is broken from how it used to work?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:68]: I think we are missing tests where the `id_token` key is not present in the body at all, that the body is totally empty

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:72]: In general, I see that the given token is expected to be in cleartext and not in base64. Will that work for our new flow we need to support?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:266]: This test also puts the token in the body. The name above implies that it will check the token from the header but that won't be the case. Am I right?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:287]: This context has the exact name as above. Can you update it to describe its true purpose?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:314]: Also in body, not just in header. Can you please update the description to be more accurate?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:341]: Same as above, can you please update this description also with the state of the body of this test?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (rafis3) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:352]: Can you also add a test that checks that the Authorization header exists, with a value of Bearer but nothing after it?
Another test that I would add, is that there is an Authorization header, and the first word is not Bearer, while the second word is a valid token (should fail)

**Review by rafis3 — CHANGES_REQUESTED** [review]
@oburstein-hub thanks, please take a look at my comments

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (oburstein-hub) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:15]: done

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (oburstein-hub) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:39]: done

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:48]: fixed

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:68]: added

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:72]: yes - they both work - I have checked

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:266]: We will check token from header. it is missing in bode in this case

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:287]: deleted it

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:314]: done

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:341]: done

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (oburstein-hub) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:352]: done

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (rafis3) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:38]: We are still left with the `class << self` block. Why do we need it?

Comment (oburstein-hub) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb]: managed to drop it now

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (jvanderhoof) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:66]: Doesn't the `Bearer` value come through as a Base64 encoded JWT?

Comment (jvanderhoof) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:66]: Just saw the comment below.  Maybe one of these is Base64 encoded to ensure the typical flow is tested?

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (jvanderhoof) [inline spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:287]: The `mocked_decode_and_verify_id_token` mock accepts and parses JSON: https://github.com/cyberark/conjur/blob/master/spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb#L18. 

It looks like this will fail if we pass Base64 encoded JSON. When the bearer token is passed in the body, is it passed as JSON?

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (jvanderhoof) [inline app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:39]: The `@` denotes an instance variable, which means it's available outside the method. Unless we need `@id_token` outside this method, please use the local variable `id_token`.

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (jvanderhoof) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:15]: @oburstein-hub, avoiding a cascading change feels like the right tradeoff.  Mind adding a comment as to why we're adding the `request` object (so future engineers understand the decision)?

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (aloncarmel111) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb]: fixed

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (aloncarmel111) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb]: fixed

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (aloncarmel111) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:16]: added comment

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (aloncarmel111) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:66]: The token is base64 encoded, we implemented same test as the `id_token` present in the body, will create integration test for Conjur cloud UI in different story

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (aloncarmel111) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:287]: yes, we implemented same test as the `id_token` in the body.

Comment (aloncarmel111) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:287]: yes, we implemented same test as the `id_token` in the body.

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (micahlee) [inline app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:56]: Hey @aloncarmel111 , this is some tricky logic to follow, what do you think about something like this instead?

```ruby
      # The credentials are in a URL encoded form data in the request body or in the request header
      def decoded_credentials
        @decoded_credentials ||= begin
          return token_from_header if token_from_header

          return token_from_body if token_from_body

          # If the token is not in the header or body, raise a missing param exception
          raise Errors::Authentication::RequestBody::MissingRequestParam, 'id_token'
        end
      end

      def token_from_header
        @token_from_header ||= begin
          return nil unless request&.headers&.key?("HTTP_AUTHORIZATION")

          # Extract the token from the authorization header
          id_token = request.headers['HTTP_AUTHORIZATION'].split(' ', -1)

          # Verify the token is given as a bearer token
          return nil unless id_token[0] == "Bearer" && id_token.length == 2

          # Return the token formatted as it would be if extracted from the body
          # as `application/x-www-form-urlencoded` data.
          {
            # URL decode the ID token from the header
            "id_token" => URI.decode_www_form_component(id_token[1])
          }
        end
      end

      def token_from_body
        @token_from_body ||= begin
          # Parse the request body
          credential_token = Hash[URI.decode_www_form(credentials)]

          # Return the parsed body if it include the token
          credential_token unless credential_token.fetch('id_token', '').empty?
        end
      end
```

**Review by micahlee — COMMENTED** [review]
Hi @aloncarmel111 , most of this looks good, thanks! I had one suggestion for trying to make the logic a little easier to read/follower. Let me know what you think!

**Comment by micahlee** [comment]
Thanks @aloncarmel111 , the only blocker I have left is to squash the PR review commits into the one `Support authorization token in header for OIDC authentication` commit before approving. 

It looks like there are still unresolved discussions around base64 encoding the header in the tests with Jason and Rafi. So you can wait to do that squash until that is resolved, and then ping me for approval. Thanks!

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (jvanderhoof) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:59]: I like the move to break up these methods to better cover edge cases.  I'd suggest passing the header in as an argument:

```suggestion
      def token_from_header(header)
```

Passing the header into the method will greatly simplify writing tests (which it looks like need to be written).

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (jvanderhoof) [inline app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:78]: Same suggestion as above, maybe move to something like:

```ruby
token_from_body(Hash[URI.decode_www_form(credentials)]

def token_from_body(credential_token)
  # Return the parsed body if it include the token
  credential_token unless credential_token.fetch('id_token', "").empty?
end
```

The above change makes the method stateless and should simplify testing the happy and sad paths.

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (jvanderhoof) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:66]: Please add at least a unit test here as well. I'm not comfortable adding functionality without tests to back it up.  The Core team will be on the line for supporting this work.

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (micahlee) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:59]: @jvanderhoof 
> I'd suggest passing the header in as an argument:

These are internal private methods, so we wouldn't expect them to have unit tests for themselves. Rather, `request` is the input that would need to be manipulated for the public interface of `UpdateInputWithUsernameFromIdToken`. Adding parameters to these methods means we're passing internal class state between internal class methods. No?

If we want to unit tests these separately, that should be factored out and injected as dependencies here instead. I don't know that that's warranted in this case, but I don't think we should add parameters to them as class methods now.

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (aloncarmel111) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:66]: added integration test

File: app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb
Comment (jvanderhoof) [reply app/domain/authentication/authn_oidc/update_input_with_username_from_id_token.rb:59]: Thanks @micahlee. I'm ok moving forward with what we have here.

File: cucumber/authenticators_oidc/features/support/authn_oidc_helper.rb
Comment (jvanderhoof) [reply cucumber/authenticators_oidc/features/support/authn_oidc_helper.rb:31]: Never mind.  I just found the JWT generation here: https://github.com/cyberark/conjur/blob/0d12713fe96fc83a2297f5a651f4f0e70b8a0bcd/cucumber/authenticators_oidc/features/support/authn_oidc_helper.rb#L100

File: spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb
Comment (jvanderhoof) [reply spec/app/domain/authentication/authn-oidc/update_input_with_username_from_id_token_spec.rb:66]: Thanks!

**Review by jvanderhoof — APPROVED** [review]
Great work. Please rebase and cleanup commit history (ideally a single commit).

**Review by aloncarmel111 — DISMISSED** [review]

**Review by micahlee — DISMISSED** [review]

**Review by oburstein-hub — COMMENTED** [review]
Looks good

**Review by aloncarmel111 — APPROVED** [review]

**Comment by aloncarmel111** [comment]
@szh they are false positive

**Pull request merged**
